### ISSUE TITLE
no-extra-parens: Allow brackets in expressions that depend on operator precedence

### ISSUE THREAD

**Issue by dvberzon** [eslint/eslint#12682]
**What rule do you want to change?**
no-extra-parens

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Either new option or new default behaviour. Option may be less contraversial

**Please provide some example code that this change will affect:**

```js
const cost = fixed + (monthly * 12);
```

**What does the rule currently do for this code?**
Warn about the "Unnecessary" parentheses

**What will the rule do after it's changed?**
No warning

**Are you willing to submit a pull request to implement this change?**
I could try.

The justification for this change comes from the results of our readability experiment that looks for empirical data on how coding constructs affect readability. The code above was included in a series of blind A/B tests where we measured the time taken by developers to read a code snippet and whether they could correctly predict its outcome.

We found a statistically significant improvement in both speed of reading the code and accuracty of prediction when the parentheses were included in the code. Whilst they are technically un-necessary if operator precedence is followed, it seems clear that adding them improves readability.

The full results of the experiment are available here:

https://howreadable.netlify.com/results/#operator-precedence

**Comment by mdjermanovic** [comment]
Hi @dvberzon, thanks for the very interesting and useful analysis!

Does the existing `nestedBinaryExpressions` option cover all the specific recommendations about parentheses? It allows this particular example:

```js
/* eslint no-extra-parens: ["error", "all", { "nestedBinaryExpressions": false }] */

const cost = fixed + (monthly * 12); // no error
```

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWV4dHJhLXBhcmVuczogW1wiZXJyb3JcIiwgXCJhbGxcIiwgeyBcIm5lc3RlZEJpbmFyeUV4cHJlc3Npb25zXCI6IGZhbHNlIH1dICovXG5cbmNvbnN0IGNvc3QgPSBmaXhlZCArIChtb250aGx5ICogMTIpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Comment by dvberzon** [comment]
The option does allow for all of the snippets we tested

[demo with our snippets](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWV4dHJhLXBhcmVuczogW1wiZXJyb3JcIiwgXCJhbGxcIiwgeyBcIm5lc3RlZEJpbmFyeUV4cHJlc3Npb25zXCI6IGZhbHNlIH1dICovXG5cbmNvbnN0IGNvc3QgPSBmaXhlZCArIChtb250aGx5ICogMTIpO1xuXG5jb25zdCByZXN1bHQgPSAodHJ1ZSAmJiB0cnVlKSB8fCAodHJ1ZSAmJiBmYWxzZSk7XG5cbmNvbnN0IGFnZUJvb2wgPSAoYWdlIDwgMjApICYmIChhZ2UgPiAxMik7XG5cbmNvbnN0IHNhbGFyeUJvb2wgPSBzYWxhcnkgLSAoc2FsYXJ5ICogdGF4UmF0ZSkgKyBib251cztcblxuY29uc3Qgc3RyaW5nVmFsID0gKHByZWZpeCB8fCAnJCcpICsgdmFsdWU7XG5cbmNvbnN0IGVtcGxveWVlcyA9IGFwcGxpY2F0aW9ucyArIChlbXBsb3llZXMgKiBlbXBsb3llZVR1cm5vdmVyKTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

Thank you for pointing that option out. I'm confused by the naming of the option, as I'm not sure how the examples shown are binary expressions:

https://eslint.org/docs/rules/no-extra-parens#nestedbinaryexpressions

However, this option covers the cases we tested. Perhaps it's worth considering having the option as false by default?

Thank you for your consideration of our results.

**Comment by mdjermanovic** [comment]
> Thank you for pointing that option out. I'm confused by the naming of the option, as I'm not sure how the examples shown are binary expressions:
> 
> https://eslint.org/docs/rules/no-extra-parens#nestedbinaryexpressions

It's 'binary' as - expression with a binary operator (that requires two operands, unlike unary and ternary operators). I guess it could be confused with bitwise operators which work on the binary representation of the data? 

 > Perhaps it's worth considering having the option as false by default?

It makes sense to me, as the default settings can be seen as a recommendation, but that would be a breaking change and it might be confusing because the rule by default disallows all parens (except those that are the responsibility of other rules).

**Comment by mdjermanovic** [comment]
Just to add, ESLint has also a core rule to basically enforce these parentheses: [no-mixed-operators](https://eslint.org/docs/rules/no-mixed-operators)
